README.md
<h1 align="center">
  <a href="https://umi.top"><img src="https://umi-top.github.io/logo.svg" alt="UMI" width="200"></a>
  <br>
  UMI Core - JavaScript Library
  <br>
  <br>
</h1>

<p align="center">
  <!-- build      --><a href="https://travis-ci.org/umi-top/umi-core-js"><img alt="travis" src="https://img.shields.io/travis/umi-top/umi-core-js/master"></a>
  <!-- coverage   --><img alt="Coveralls github branch" src="https://img.shields.io/coveralls/github/umi-top/umi-core-js/master">
  <!-- code style --><a href="https://standardjs.com"><img alt="Standard" src="https://img.shields.io/badge/code_style-standard-green"></a>
  <!-- license    --><a href="https://github.com/umi-top/umi-core-js/blob/master/LICENSE"><img alt="GitHub" src="https://img.shields.io/github/license/umi-top/umi-core-js"></a>
  <!-- PGP        --><a href="https://keybase.io/umitop"><img alt="Keybase PGP" src="https://img.shields.io/keybase/pgp/umitop"></a>
  <br/>
  <!-- release    --><a href="https://github.com/umi-top/umi-core-js"><img alt="GitHub release (latest SemVer)" src="https://img.shields.io/github/v/release/umi-top/umi-core-js?sort=semver"></a>
  <!-- npm ver    --><a href="https://www.npmjs.com/package/@umi-top/umi-core-js"><img alt="npm version" src="https://img.shields.io/npm/v/@umi-top/umi-core-js"></a>
  <!-- types      --><img alt="npm type definitions" src="https://img.shields.io/npm/types/@umi-top/umi-core-js">
  <!-- node       --><a href="https://www.npmjs.com/package/@umi-top/umi-core-js"><img alt="node-current (scoped)" src="https://img.shields.io/node/v/@umi-top/umi-core-js"></a>
  <!-- downloads  --><a href="https://www.npmjs.com/package/@umi-top/umi-core-js"><img alt="npm downloads" src="https://img.shields.io/npm/dm/@umi-top/umi-core-js"></a>
</p>

## Оглавление
- [Введение](#введение)
- [Установка](#установка)
  - [npm](#npm)
  - [yarn](#yarn)
  - [Отдельным файлом](#отдельным-файлом)
- [Подключение](#подключение)
  - [CommonJS](#commonjs)
  - [ES Modules](#es-modules)
  - [CDN](#cdn)
- [Примеры](#примеры)
  - [Мнемоники](#мнемоники)
    - [Seed из мнемонической фразы](#seed-из-мнемонической-фразы)
  - [Ключи](#ключи)
    - [Ключ из seed](#ключ-из-seed)
    - [Подписать сообщение](#подписать-сообщение)
    - [Проверить подпись](#проверить-подпись)
  - [Адреса](#адреса)
    - [Адрес в формате Bech32](#адрес-в-формате-bech32)
    - [Адрес из приватного или публичного ключа](#адрес-из-приватного-или-публичного-ключа)
    - [Префикс](#префикс)
  - [Транзакции](#транзакции)
    - [Перевести монеты](#перевести-монеты)
    - [Создать структуру](#создать-структуру)
    - [Обновить настройки структуры](#обновить-настройки-структуры)
    - [Установить адрес для начисления профита](#установить-адрес-для-начисления-профита)
    - [Установить адрес для перевода комиссии](#установить-адрес-для-перевода-комиссии)
    - [Активировать транзитный адрес](#активировать-транзитный-адрес)
    - [Деактивировать транзитный адрес](#деактивировать-транзитный-адрес)
    - [Отправить транзакцию в сеть](#отправить-транзакцию-в-сеть)
- [Лицензия](#лицензия)

## Введение

Библиотека написана на [TypeScript](https://www.typescriptlang.org) и
скомпилирована в модули CommonJS и ES Module и работает
во всех актуальных браузерах и Node.js.  
Для совместимости с устаревшими браузерами скомпилирована iife версия,
поддерживается Internet Explorer 5 и выше.

## Установка

Библиотека опубликована в репозитории
[npm](https://www.npmjs.com/package/@umi-top/umi-core-js) и может быть
установлена с помощью пакетного менеджера, например,
[npm](https://docs.npmjs.com) или [Yarn](https://yarnpkg.com):

### npm
```bash
npm install @umi-top/umi-core-js
```
### yarn
```bash
yarn add @umi-top/umi-core-js
```

### Отдельным файлом

Если требуется, библиотеку можно скачать в виде отдельного файла:  
CommonJS: [index.js](https://unpkg.com/@umi-top/umi-core-js/lib/index.js),
ES Module: [index.mjs](https://unpkg.com/@umi-top/umi-core-js/lib/index.mjs)
или IIFE [index.min.js](https://unpkg.com/@umi-top/umi-core-js/lib/index.min.js).  
Так же можно скачать аннотации типов для
TypeScript: [index.d.ts](https://unpkg.com/@umi-top/umi-core-js/lib/index.d.ts)
и Flow: [index.js.flow](https://unpkg.com/@umi-top/umi-core-js/lib/index.js.flow).

## Подключение

### CommonJS

Если вы используете [Node.js](https://nodejs.org/api/modules.html):

```javascript
const umi = require('@umi-top/umi-core-js')
```

### ES Modules

Если вы используете [webpack](https://webpack.js.org/guides/getting-started/#modules),
[Rollup](https://rollupjs.org/guide/en/#importing),
[Parcel](https://parceljs.org/javascript.html) или [Node.js (>= v13)](https://nodejs.org/api/esm.html).

Для более эффективной работы алгоритма [tree shaking](https://webpack.js.org/guides/tree-shaking/)
рекомендуем импортировать только используемые значения:

```javascript
import { Address, PublicKey, SecretKey, Transaction } from '@umi-top/umi-core-js'
```

Импортировать все содержимое модуля.
Этот вариант будет использоваться в примерах ниже.

```javascript
import * as umi from '@umi-top/umi-core-js'
```

### CDN

Библиотеку можно использовать напрямую в браузере.

Если требуется максимальная совместимость, можно использовать
[IIFE](https://developer.mozilla.org/en-US/docs/Glossary/IIFE) вариант:

```html
<script src="https://unpkg.com/@umi-top/umi-core-js"></script>
```

В [современных](https://caniuse.com/#feat=es6-module) браузерах можно
импортировать [es6 модуль](https://v8.dev/features/modules):

```html
<script type="module">
import * as umi from 'https://unpkg.com/@umi-top/umi-core-js/lib/index.mjs'
</script>
```

## Примеры

### Мнемоники

UMI не накладывает никаких ограничений на способ генерации и хранения приватных
ключей.  
Для совместимости, рекомендуем использовать
[bip39](https://github.com/bitcoin/bips/blob/master/bip-0039.mediawiki).

#### Seed из мнемонической фразы

Пример с использованием библиотеки [bip39](https://www.npmjs.com/package/bip39):

```javascript
// npm install bip39

const bip39 = require('bip39')
const mnemonic = bip39.generateMnemonic(256)
const seed = bip39.mnemonicToSeedSync(mnemonic)
```

### Ключи

В UMI применяется [Ed25519](https://ed25519.cr.yp.to)
([RFC 8032](https://tools.ietf.org/html/rfc8032)) —
схема подписи [EdDSA](https://ru.wikipedia.org/wiki/EdDSA) использующая
[SHA-512](https://en.wikipedia.org/wiki/SHA-2)
и [Curve25519](https://en.wikipedia.org/wiki/Curve25519). 

#### Ключ из seed

Seed может быть любой длины. Оптимальным вариантом является длина 32 байта.

```javascript
const seed = new Uint8Array(32)
const secretKey = umi.SecretKey.fromSeed(seed)
const publicKey = secretKey.getPublicKey()
```

#### Подписать сообщение

В метод `SecretKey#sign()` необходимо передать массив байтов, поэтому если
требуется подписать текстовое сообщение его нужно преобразовать:

```javascript
const message = umi.textEncode('Hello World')
const signature = secretKey.sign(message)

console.log(umi.base64Encode(signature))
```

#### Проверить подпись

Метод `PublicKey#verifySignature()` принимает массив байтов, поэтому если
подпись передается в виде текста ее необходимо декодировать:

```javascript
const address = 'umi18d4z00xwk6jz6c4r4rgz5mcdwdjny9thrh3y8f36cpy2rz6emg5s6rxnf6'
const message = umi.textEncode('Hello World')
const signature = umi.base64Decode(
  'Jbi9YfwLcxiTMednl/wTvnSzsPP9mV9Bf2vvZytP87oyg1p1c9ZBkn4gNv15ZHwEFv3bVYlowgyIKmMwJLjJCw=='
)
const pubKey = umi.Address.fromBech32(address).getPublicKey()
const isValid = pubKey.verifySignature(signature, message)

console.log(isValid ? 'true' : 'false')
```

### Адреса

UMI использует адреса в формате Bech32
([bip173](https://github.com/bitcoin/bips/blob/master/bip-0173.mediawiki))
длиной 62 символа. Префикс имеет длину 3 символа.  
Специальным случаем являются Genesis-адреса, существующие только
в Genesis-блоке, такие адреса имеют длину 65 символов
и всегда имеют префикс `genesis`.

#### Адрес в формате Bech32

Создать адрес из строки Bech32 можно используя статический метод `Address.fromBech32()`
и экспортировать с помощью `Address#getBech32()`:

```javascript
const bech32 = 'umi18d4z00xwk6jz6c4r4rgz5mcdwdjny9thrh3y8f36cpy2rz6emg5s6rxnf6'
const address = umi.Address.fromBech32(bech32)

console.log(address.getBech32())
```
 
#### Адрес из приватного или публичного ключа

Статический метод `Address.fromKey()` создает адрес из приватного
или публичного ключа:

```javascript
const seed = new Uint8Array(32)
const secKey = umi.SecretKey.fromSeed(seed)
const pubKey = secKey.getPublicKey()
const address1 = umi.Address.fromKey(secKey)
const address2 = umi.Address.fromKey(pubKey)

console.log(address1.getBech32(), address2.getBech32())
```

#### Префикс

По умолчанию адреса имеют префикс `umi`.
Изменить префикс можно с помощью метода `Address#setPrefix()`:

```javascript
const bech32 = 'umi18d4z00xwk6jz6c4r4rgz5mcdwdjny9thrh3y8f36cpy2rz6emg5s6rxnf6'
const address = umi.Address.fromBech32(bech32).setPrefix('aaa')

console.log(address.getBech32())
```

### Транзакции

#### Перевести монеты

Поля `sender`, `recipient` и `value` являются обязательными.  
Сумма указывается в UMI-центах, т.е. 1.23 UMI = 123.

```javascript
const secKey = umi.SecretKey.fromSeed(new Uint8Array(32))
const sender = umi.Address.fromKey(secKey).setPrefix('umi')
const recipient = umi.Address.fromKey(secKey).setPrefix('aaa')
const tx = new umi.Transaction()
  .setVersion(umi.Transaction.Basic)
  .setSender(sender)
  .setRecipient(recipient)
  .setValue(42)
  .sign(secKey)

console.log('txid:   ', umi.hexEncode(tx.getHash()))
console.log('base64: ', umi.base64Encode(tx.getBytes()))
```

#### Создать структуру

Поля `sender`, `prefix`, `name`, `profitPercent` и `feePercent` являются
обязательными.  
Префикс имеет длину 3 символа. Название указывается в кодировке UTF-8 и может
иметь длину до 35 байтов. Проценты указываются в сотых долях процента,
т.е. 1.23% = 123.

```javascript
const secKey = umi.SecretKey.fromSeed(new Uint8Array(32))
const sender = umi.Address.fromKey(secKey).setPrefix('umi')
const tx = new umi.Transaction()
  .setVersion(umi.Transaction.CreateStructure)
  .setSender(sender)
  .setPrefix('aaa')
  .setName('🙂')
  .setProfitPercent(500)
  .setFeePercent(2000)
  .sign(secKey)

console.log('txid:   ', umi.hexEncode(tx.getHash()))
console.log('base64: ', umi.base64Encode(tx.getBytes()))
```

#### Обновить настройки структуры

Поля `sender`, `prefix`, `name`, `profitPercent` и `feePercent` являются
обязательными.  
Необходимо задать все поля, даже если они не изменились.

```javascript
const secKey = umi.SecretKey.fromSeed(new Uint8Array(32))
const sender = umi.Address.fromKey(secKey).setPrefix('umi')
const tx = new umi.Transaction()
  .setVersion(umi.Transaction.UpdateStructure)
  .setSender(sender)
  .setPrefix('aaa')
  .setName('🙂')
  .setProfitPercent(500)
  .setFeePercent(2000)
  .sign(secKey)

console.log('txid:   ', umi.hexEncode(tx.getHash()))
console.log('base64: ', umi.base64Encode(tx.getBytes()))
```

#### Установить адрес для начисления профита

Поля `sender` и `recipient` являются обязательными.  
Адрес для начисления профита должен принадлежать структуре.

```javascript
const secKey = umi.SecretKey.fromSeed(new Uint8Array(32))
const sender = umi.Address.fromKey(secKey).setPrefix('umi')
const newPrf = umi.Address.fromBech32('aaa18d4z00xwk6jz6c4r4rgz5mcdwdjny9thrh3y8f36cpy2rz6emg5svsuw66')
const tx = new umi.Transaction()
  .setVersion(umi.Transaction.UpdateProfitAddress)
  .setSender(sender)
  .setRecipient(newPrf)
  .sign(secKey)

console.log('txid:   ', umi.hexEncode(tx.getHash()))
console.log('base64: ', umi.base64Encode(tx.getBytes()))
```

#### Установить адрес для перевода комиссии

Поля `sender` и `recipient` являются обязательными.  
Адрес для перевода комиссии должен принадлежать структуре.

```javascript
const secKey = umi.SecretKey.fromSeed(new Uint8Array(32))
const sender = umi.Address.fromKey(secKey).setPrefix('umi')
const newFee = umi.Address.fromBech32('aaa18d4z00xwk6jz6c4r4rgz5mcdwdjny9thrh3y8f36cpy2rz6emg5svsuw66')
const tx = new umi.Transaction()
  .setVersion(umi.Transaction.UpdateFeeAddress)

CHANGELOG.md
# Changelog

## [1.0.0][] - 2020-06-14

### Added

- Классы `SecretKey`, `PublicKey`, `Address`, `Transaction`
- Функции `textEncode()`/`textDecode()`, `base64Encode()`/`base64Decode()`, `hexEncode()`/`hexDecode()`

[1.0.0]: https://github.com/umi-top/umi-core-js/releases/tag/v1.0.0

lib/index.d.ts
/**
 * Класс для работы с адресами.
 * @class
 */
export declare class Address {
    /**
     * @example
     * let address = new Address()
     */
    constructor();
    /**
     * Статический метод, создает объект из адреса в формате Bech32.
     * @param {string} bech32 Адрес в формате Bech32, длина 62 или 65 символов.
     * @returns {Address}
     * @throws {Error}
     * @example
     * let address = Address.fromBech32('umi18d4z00xwk6jz6c4r4rgz5mcdwdjny9thrh3y8f36cpy2rz6emg5s6rxnf6')
     */
    static fromBech32(bech32: string): Address;
    /**
     * Статический метод, создает объект из бинарного представления.
     * @param {ArrayLike<number>} bytes Адрес в бинарном виде, длина 34 байта.
     * @returns {Address}
     * @throws {Error}
     * @example
     * let address = Address.fromBytes(new Uint8Array(34))
     */
    static fromBytes(bytes: ArrayLike<number>): Address;
    /**
     * Статический метод, создает объект из публичного или приватного ключа.
     * @param {(PublicKey|SecretKey)} key Публичный или приватный ключ.
     * @returns {Address}
     * @example
     * let secKey = SecretKey.fromSeed([])
     * let address = Address.fromKey(secKey)
     */
    static fromKey(key: PublicKey | SecretKey): Address;
    /**
     * Адрес в формате Bech32, длина 62 или 65 символов.
     * @returns {string}
     * @example
     * let bech32 = new Address().getBech32()
     */
    getBech32(): string;
    /**
     * Адрес в бинарном виде, длина 34 байта.
     * @returns {number[]}
     * @example
     * let bytes = new Address().getBytes()
     */
    getBytes(): number[];
    /**
     * Префикс адреса, три символа латиницы в нижнем регистре.
     * @returns {string}
     * @throws {Error}
     * @example
     * let prefix = new Address().getPrefix()
     */
    getPrefix(): string;
    /**
     * Устанавливает префикс адреса и возвращает this.
     * @param {string} prefix Префикс адреса, три символа латиницы в нижнем регистре.
     * @returns {Address}
     * @throws {Error}
     * @example
     * let address = new Address().setPrefix('umi')
     */
    setPrefix(prefix: string): Address;
    /**
     * Публичный ключ.
     * @returns {PublicKey}
     * @example
     * let pubKey = new Address().getPublicKey()
     */
    getPublicKey(): PublicKey;
    /**
     * Устанавливает публичный ключ и возвращает this.
     * @param {PublicKey} publicKey Публичный ключ.
     * @returns {Address}
     * @throws {Error}
     * @example
     * let pubKey = SecretKey.fromSeed([]).getPublicKey()
     * let address = new Address().setPublicKey(pubKey)
     */
    setPublicKey(publicKey: PublicKey): Address;
}
/**
 * Класс для работы с публичными ключами.
 * @class
 */
export declare class PublicKey {
    /**
     * @param {ArrayLike<number>} bytes Публичный ключ в формате libsodium, 32 байта.
     * @throws {Error}
     * @example
     * let bytes = new Uint8Array(32)
     * let pubKey = new PublicKey(bytes)
     */
    constructor(bytes: ArrayLike<number>);
    /**
     * Публичный ключ в формате libsodium, 32 байта.
     * @returns {number[]}
     * @example
     * let bytes = new PublicKey(new Uint8Array(32)).getBytes()
     */
    getBytes(): number[];
    /**
     * Проверяет цифровую подпись.
     * @param {ArrayLike<number>} signature Подпись, 64 байта.
     * @param {ArrayLike<number>} message Сообщение.
     * @returns {boolean}
     * @throws {Error}
     * @example
     * let pubKey = new PublicKey(new Uint8Array(32))
     * let signature = new Uint8Array(64)
     * let message = new TextEncoder().encode('Hello World')
     * let isValid = pubKey.verifySignature(signature, message)
     */
    verifySignature(signature: ArrayLike<number>, message: ArrayLike<number>): boolean;
}
/**
 * Класс для работы с приватными ключами.
 * @class
 */
export declare class SecretKey {
    /**
     * @param {ArrayLike<number>} bytes Приватный ключ в бинарном виде.
     * В формате libsodium, 64 байта.
     * @throws {Error}
     * @example
     * let bytes = SecretKey.fromSeed(new Uint8Array(32)).getBytes()
     * let secKey = new SecretKey(bytes)
     */
    constructor(bytes: ArrayLike<number>);
    /**
     * Статический метод, создающий приватный ключ из seed.\
     * Libsodium принимает seed длиной 32 байта, поэтому если длина
     * отличается, то берется sha256 хэш.
     * @param {ArrayLike<number>} seed Массив байтов любой длины.
     * @returns {SecretKey}
     * @example
     * let seed = new Uint8Array(32)
     * let secKey = SecretKey.fromSeed(seed)
     */
    static fromSeed(seed: ArrayLike<number>): SecretKey;
    /**
     * Приватный ключ в бинарном виде. В формате libsodium, 64 байта.
     * @returns {number[]}
     * @example
     * let secKey = SecretKey.fromSeed(new Uint8Array(32))
     * let bytes = secKey.getBytes()
     */
    getBytes(): number[];
    /**
     * Публичный ключ, соответствующий приватному ключу.
     * @returns {PublicKey}
     * @example
     * let secKey = SecretKey.fromSeed(new Uint8Array(32))
     * let pubKey = secKey.getPublicKey()
     */
    getPublicKey(): PublicKey;
    /**
     * Создает цифровую подпись сообщения.
     * @param {ArrayLike<number>} message Сообщение, которое необходимо подписать.
     * @returns {number[]} Подпись длиной 64 байта.
     * @example
     * let secKey = SecretKey.fromSeed(new Uint8Array(32))
     * let message = new TextEncoder().encode('Hello World')
     * let signature = secKey.sign(message)
     */
    sign(message: ArrayLike<number>): number[];
}
/**
 * Класс для работы с транзакциями.
 * @class
 */
export declare class Transaction {
    /**
     * Genesis-транзакция.\
     * Может быть добавлена только в Genesis-блок.\
     * Адрес отправителя должен иметь префикс genesis, адрес получателя - umi.
     * @type {number}
     * @constant
     * @example
     * let secKey = SecretKey.fromSeed(new Uint8Array(32))
     * let sender = Address.fromKey(secKey).setPrefix('genesis')
     * let recipient = Address.fromKey(secKey).setPrefix('umi')
     * let tx = new Transaction()
     *   .setVersion(Transaction.Genesis)
     *   .setSender(sender)
     *   .setRecipient(recipient)
     *   .setValue(42)
     *   .sign(secKey)
     */
    static Genesis: number;
    /**
     * Стандартная транзакция. Перевод монет из одного кошелька в другой.
     * @type {number}
     * @constant
     * @example
     * let secKey = SecretKey.fromSeed(new Uint8Array(32))
     * let sender = Address.fromKey(secKey).setPrefix('umi')
     * let recipient = Address.fromKey(secKey).setPrefix('aaa')
     * let tx = new Transaction()
     *   .setVersion(Transaction.Basic)
     *   .setSender(sender)
     *   .setRecipient(recipient)
     *   .setValue(42)
     *   .sign(secKey)
     */
    static Basic: number;
    /**
     * Создание новой структуры.
     * @type {number}
     * @constant
     * @example
     * let secKey = SecretKey.fromSeed(new Uint8Array(32))
     * let sender = Address.fromKey(secKey).setPrefix('umi')
     * let tx = new Transaction()
     *   .setVersion(Transaction.CreateStructure)
     *   .setSender(sender)
     *   .setPrefix('aaa')
     *   .setName('My Struct 🙂')
     *   .setProfitPercent(100)
     *   .setFeePercent(0)
     *   .sign(secKey)
     */
    static CreateStructure: number;
    /**
     * Обновление настроек существующей структуры.
     * @type {number}
     * @constant
     * @example
     * let secKey = SecretKey.fromSeed(new Uint8Array(32))
     * let sender = Address.fromKey(secKey).setPrefix('umi')
     * let tx = new Transaction()
     *   .setVersion(Transaction.UpdateStructure)
     *   .setSender(sender)
     *   .setPrefix('aaa')
     *   .setName('My New Struct 😎')
     *   .setProfitPercent(500)
     *   .setFeePercent(2000)
     *   .sign(secKey)
     */
    static UpdateStructure: number;
    /**
     * Изменение адреса для начисления профита.
     * @type {number}
     * @constant
     * @example
     * let secKey = SecretKey.fromSeed(new Uint8Array(32))
     * let sender = Address.fromKey(secKey).setPrefix('umi')
     * let newPrf = Address.fromBech32('aaa18d4z00xwk6jz6c4r4rgz5mcdwdjny9thrh3y8f36cpy2rz6emg5svsuw66')
     * let tx = new Transaction()
     *   .setVersion(Transaction.UpdateProfitAddress)
     *   .setSender(sender)
     *   .setRecipient(newPrf)
     *   .sign(secKey)
     */
    static UpdateProfitAddress: number;
    /**
     * Изменение адреса на который переводится комиссия.
     * @type {number}
     * @constant
     * @example
     * let secKey = SecretKey.fromSeed(new Uint8Array(32))
     * let sender = Address.fromKey(secKey).setPrefix('umi')
     * let newFee = Address.fromBech32('aaa18d4z00xwk6jz6c4r4rgz5mcdwdjny9thrh3y8f36cpy2rz6emg5svsuw66')
     * let tx = new Transaction()
     *   .setVersion(Transaction.UpdateFeeAddress)
     *   .setSender(sender)
     *   .setRecipient(newFee)
     *   .sign(secKey)
     */
    static UpdateFeeAddress: number;
    /**
     * Активация транзитного адреса.
     * @type {number}
     * @constant
     * @example
     * let secKey = SecretKey.fromSeed(new Uint8Array(32))
     * let sender = Address.fromKey(secKey).setPrefix('umi')
     * let transit = Address.fromBech32('aaa18d4z00xwk6jz6c4r4rgz5mcdwdjny9thrh3y8f36cpy2rz6emg5svsuw66')
     * let tx = new Transaction()
     *   .setVersion(Transaction.CreateTransitAddress)
     *   .setSender(sender)
     *   .setRecipient(transit)
     *   .sign(secKey)
     */
    static CreateTransitAddress: number;
    /**
     * Деактивация транзитного адреса.
     * @type {number}
     * @constant
     * @example
     * let secKey = SecretKey.fromSeed(new Uint8Array(32))
     * let sender = Address.fromKey(secKey).setPrefix('umi')
     * let transit = Address.fromBech32('aaa18d4z00xwk6jz6c4r4rgz5mcdwdjny9thrh3y8f36cpy2rz6emg5svsuw66')
     * let tx = new Transaction()
     *   .setVersion(Transaction.DeleteTransitAddress)
     *   .setSender(sender)
     *   .setRecipient(transit)
     *   .sign(secKey)
     */
    static DeleteTransitAddress: number;
    /**
     * @example
     * let trx = new Transaction()
     */
    constructor();
    /**
     * Статический метод, создает объект из массива байтов.
     * @param {ArrayLike<number>} bytes Транзакция в бинарном виде.
     * @returns {Transaction}
     * @throws {Error}
     * @example
     * let bytes = new Uint8Array(150)
     * let trx = Transaction.fromBytes(bytes)
     */
    static fromBytes(bytes: ArrayLike<number>): Transaction;
    /**
     * Транзакция в бинарном виде, 150 байт.
     * @returns {number[]}
     * @example
     * let bytes = new Transaction().getBytes()
     */
    getBytes(): number[];
    /**
     * Хэш (txid) транзакции.
     * @returns {number[]}
     * @example
     * let hash = new Transaction().getHash()
     */
    getHash(): number[];
    /**
     * Версия (тип) транзакции.
     * @returns {number}
     * @example
     * let ver = new Transaction().getVersion()
     */
    getVersion(): number;
    /**
     * Устанавливает версию и возвращает this.
     * @param {number} version Версия (тип) транзакции.
     * @returns {Transaction}
     * @throws {Error}
     * @see Transaction.Genesis
     * @see Transaction.Basic
     * @see Transaction.CreateStructure
     * @see Transaction.UpdateStructure
     * @see Transaction.UpdateProfitAddress
     * @see Transaction.UpdateFeeAddress
     * @see Transaction.CreateTransitAddress
     * @see Transaction.DeleteTransitAddress
     */
    setVersion(version: number): Transaction;
    /**
     * Отправитель.\
     * Доступно для всех типов транзакций.
     * @returns {Address}
     * @example
     * let sender = new Transaction().getSender()
     */
    getSender(): Address;
    /**
     * Устанавливает отправителя и возвращает this.
     * @param {Address} address Адрес отправителя.
     * @returns {Transaction}
     * @throws {Error}
     * @example
     * let sender = new Address()
     * let trx = new Transaction().setSender(sender)
     */
    setSender(address: Address): Transaction;
    /**
     * Получатель.\
     * Недоступно для транзакций CreateStructure и UpdateStructure.
     * @returns {Address}
     * @example
     * let recipient = new Transaction().getRecipient()
     */
    getRecipient(): Address;
    /**
     * Устанавливает получателя и возвращает this.\
     * Недоступно для транзакций CreateStructure и UpdateStructure.
     * @param {Address} address Адрес получателя.
     * @returns {Transaction}
     * @throws {Error}
     * @example
     * let recipient = new Address()
     * let trx = new Transaction().setRecipient(recipient)
     */
    setRecipient(address: Address): Transaction;
    /**
     * Сумма перевода в UMI-центах, цело число в промежутке от 1 до 18446744073709551615.\
     * Доступно только для Genesis и Basic транзакций.
     * @returns {number}
     * @example
     * let value = new Transaction().getValue()
     */
    getValue(): number;
    /**
     * Устанавливает сумму и возвращает this.\
     * Принимает значения в промежутке от 1 до 18446744073709551615.\
     * Доступно только для Genesis и Basic транзакций.
     * @param {number} value Целое число от 1 до 18446744073709551615.
     * @returns {Transaction}
     * @throws {Error}
     * @example
     * let trx = new Transaction().setValue(42)
     */
    setValue(value: number): Transaction;
    /**
     * Nonce, целое число в промежутке от 0 до 18446744073709551615.\
     * Генерируется автоматически при вызове sign().
     * @returns {number}
     * @example
     * let nonce = new Transaction().getNonce()
     */
    getNonce(): number;
    /**
     * Устанавливает nonce и возвращает this.
     * @param {number} nonce Целое число в промежутке от 0 до 18446744073709551615.
     * @returns {Transaction}
     * @throws {Error}
     * @example
     * let nonce = Date.now()
     * let trx = new Transaction().setNonce(nonce)
     */
    setNonce(nonce: number): Transaction;
    /**
     * Цифровая подпись транзакции, длина 64 байта.
     * @returns {number[]}
     * @example
     * let signature = new Transaction().getSignature()
     */
    getSignature(): number[];
    /**
     * Устанавливает цифровую подпись и возвращает this.
     * @param {ArrayLike<number>} signature Подпись, длина 64 байта.
     * @returns {Transaction}
     * @throws {Error}
     * @example
     * let signature = new Uint8Array(64)
     * let trx = new Transaction().setSignature(signature)
     */
    setSignature(signature: ArrayLike<number>): Transaction;
    /**
     * Подписать транзакцию приватным ключом.
     * @param {SecretKey} secretKey Приватный ключ.
     * @returns {Transaction}
     * @throws {Error}
     * @example
     * let secKey = SecretKey.fromSeed(new Uint8Array(32))
     * let trx = new Transaction().sign(secKey)
     */
    sign(secretKey: SecretKey): Transaction;
    /**
     * Префикс адресов, принадлежащих структуре.\
     * Доступно только для CreateStructure и UpdateStructure.
     * @returns {string}
     * @throws {Error}
     * @example
     * let trx = new Transaction().setVersion(Transaction.CreateStructure)
     * let prefix = trx.getPrefix()
     */
    getPrefix(): string;
    /**
     * Устанавливает префикс и возвращает this.\
     * Доступно только для CreateStructure и UpdateStructure.
     * @param {string} prefix Префикс адресов, принадлежащих структуре.
     * @returns {Transaction}
     * @throws {Error}
     * @example
     * let trx = new Transaction().setVersion(CreateStructure)
     * trx.setPrefix('aaa')
     */
    setPrefix(prefix: string): Transaction;
    /**
     * Название структуры в кодировке UTF-8.\
     * Доступно только для CreateStructure и UpdateStructure.
     * @returns {string}
     * @throws {Error}
     * @example
     * let trx = new Transaction().setVersion(Transaction.CreateStructure)
     * let name = trx.getName()
     */
    getName(): string;
    /**
     * Устанавливает название структуры и возвращает this.\
     * Доступно только для CreateStructure и UpdateStructure.
     * @param {string} name Название структуры в кодировке UTF-8.
     * @returns {Transaction}
     * @throws {Error}
     * @example
     * let trx = new Transaction().setVersion(Transaction.CreateStructure)
     * trx.setName('Hello World')
     */
    setName(name: string): Transaction;
    /**
     * Профита в сотых долях процента с шагом в 0.01%.\
     * Принимает значения от 100 до 500 (соответственно от 1% до 5%).\
     * Доступно только для CreateStructure и UpdateStructure.
     * @returns {number}
     * @example
     * let trx = new Transaction().setVersion(Transaction.CreateStructure)
     * let profit = trx.getProfitPercent(100)
     */
    getProfitPercent(): number;
    /**
     * Устанавливает процент профита и возвращает this.\
     * Доступно только для CreateStructure и UpdateStructure.
     * @param {number} percent Профит в сотых долях процента с шагом в 0.01%.
     * Принимает значения от 100 до 500 (соответственно от 1% до 5%).
     * @returns {Transaction}
     * @throws {Error}
     * @example
     * let trx = new Transaction().setVersion(Transaction.CreateStructure)
     * trx.setProfitPercent(100)
     */
    setProfitPercent(percent: number): Transaction;
    /**
     * Комиссия в сотых долях процента с шагом в 0.01%.\
     * Принимает значения от 0 до 2000 (соответственно от 0% до 20%).\
     * Доступно только для CreateStructure и UpdateStructure.
     * @returns {number}
     * @example
     * let trx = new Transaction().setVersion(Transaction.CreateStructure)
     * let fee = trx.getFeePercent()
     */
    getFeePercent(): number;
    /**
     * Устанавливает размер комиссии и возвращает this.\
     * Доступно только для CreateStructure и UpdateStructure.
     * @param {number} percent Комиссия в сотых долях процента с шагом в 0.01%. Принимает значения от 0 до 2000 (соответственно от 0% до 20%).
     * @returns {Transaction}
     * @throws {Error}
     * @example
     * let trx = new Transaction().setVersion(Transaction.CreateStructure)
     * trx.setFeePercent(100)
     */
    setFeePercent(percent: number): Transaction;
    /**
     * Проверить транзакцию на соответствие формальным правилам.
     * @returns {boolean}
     * @throws {Error}
     * @example
     * let ver = new Transaction().verify()
     */
    verify(): boolean;
}
/**
 * Декодирует Base64 строку в массив байтов.
 * @param {string} base64 Строка в кодировке Base64.
 * @returns {number[]}
 * @throws {Error}
 */
export declare function base64Decode(base64: string): number[];
/**
 * Кодирует массив байтов в Base64 строку.
 * @param {number[]} bytes Массив байтов.
 * @returns {string}
 */
export declare function base64Encode(bytes: number[]): string;
/**
 * Декодирует Base16 строку в массив байтов.
 * @param {string} hex Строка в кодировке Base16.
 * @returns {number[]}
 * @throws {Error}
 */
export declare function hexDecode(hex: string): number[];
/**
 * Кодирует массив байтов в Base16 строку.
 * @param {number[]} bytes Массив байтов.
 * @returns {string}
 */
export declare function hexEncode(bytes: number[]): string;
/**
 * Декодирует массив байтов UTF-8 в строку в кодировке UTF-16.
 * @see https://developer.mozilla.org/en-US/docs/Web/API/TextDecoder
 * @param {number[]} bytes Массив байтов UTF-8.
 * @returns {string}
 */
export declare function textDecode(bytes: number[]): string;
/**
 * Кодирует UTF-16 строку в массив байтов UTF-8.
 * @see https://developer.mozilla.org/en-US/docs/Web/API/TextEncoder
 * @param {string} text Текстовая строка в кодировке UTF-16.
 * @returns {number[]}
 */
export declare function textEncode(text: string): number[];
